store/storage/rocksdb/iterator.go

```diff
@@ -0,0 +1,129 @@
+package rocksdb
+
+import (
+	"bytes"
+
+	"cosmossdk.io/store/v2"
+	"github.com/linxGnu/grocksdb"
+)
+
+var _ store.Iterator = (*iterator)(nil)
+
+type iterator struct {
+	source             *grocksdb.Iterator
+	prefix, start, end []byte
+	reverse            bool
+	invalid            bool
+}
+
+func newRocksDBIterator(source *grocksdb.Iterator, prefix, start, end []byte, reverse bool) *iterator {
+	if reverse {
+		if end == nil {
+			source.SeekToLast()
+		} else {
+			source.Seek(end)
+
+			if source.Valid() {
+				eoaKey := copyAndFreeSlice(source.Key()) // end or after key
+				if bytes.Compare(end, eoaKey) <= 0 {
+					source.Prev()
+				}
+			} else {
+				source.SeekToLast()
+			}
+		}
+	} else {
+		if start == nil {
+			source.SeekToFirst()
+		} else {
+			source.Seek(start)
+		}
+	}
+
+	return &iterator{
+		source:  source,
+		prefix:  prefix,
+		start:   start,
+		end:     end,
+		reverse: reverse,
+		invalid: false,
+	}
+}
+
+func (itr *iterator) Domain() ([]byte, []byte) {
+	return itr.start, itr.end
+}
+
+func (itr *iterator) Valid() bool {
+	// once invalid, forever invalid
+	if itr.invalid {
+		return false
+	}
+
+	// if source has error, consider it invalid
+	if err := itr.source.Err(); err != nil {
+		itr.invalid = true
+		return false
+	}
+
+	// if source is invalid, consider it invalid
+	if !itr.source.Valid() {
+		itr.invalid = true
+		return false
+	}
+
+	// if key is at the end or past it, consider it invalid
+	start := itr.start
+	end := itr.end
+	key := copyAndFreeSlice(itr.source.Key())
```

Comment:
This does feel like it'll be expensive given that we are simply comparing key against start and end and not going to mutate it. Is it possible for us to just get a non-deep copy of itr.source.Key() just for this comparison?

Could we perhaps write a helper function
```go
func sliceForReadOnly(s *grocksdb.Slice) []byte {
   if !s.Exists() {
      return nil
   }
   return s.Data()[:]
}
```

and then use that for every single comparison given that we are going to discard it anyways.